// @graphql-mesh/merger-bare/index.js
'use strict';

const wrap = require('@graphql-tools/wrap');
const utils = require('@graphql-mesh/utils');
const schema = require('@graphql-tools/schema');

class BareMerger {
    constructor(options) {
        this.name = 'bare';
        this.logger = options.logger;
    }
    async getUnifiedSchema({ rawSources, typeDefs, resolvers, transforms }) {
        const sourceMap = new Map();
        this.logger.debug(() => `Applying transforms for each source`);
        const schemas = rawSources.map(source => {
            let schema = source.schema;
            let sourceLevelSchema = source.schema;
            const handlerLevelTransformGroups = utils.groupTransforms(source.transforms || []);
            if (handlerLevelTransformGroups.wrapTransforms.length > 0 || source.executor) {
                schema = wrap.wrapSchema({
                    batch: true,
                    ...source,
                    schema,
                });
            }
            else {
                schema = utils.applySchemaTransforms(schema, undefined, schema, handlerLevelTransformGroups.noWrapTransforms);
            }
            // After that step, it will be considered as root level schema
            sourceLevelSchema = schema;
            sourceMap.set(source, sourceLevelSchema);
            return schema;
        });
        this.logger.debug(() => `Merging sources`);
        let unifiedSchema = schemas.length === 1 && !(typeDefs === null || typeDefs === void 0 ? void 0 : typeDefs.length) && !resolvers
            ? schemas[0]
            : schema.mergeSchemas({
                schemas,
                typeDefs,
                resolvers,
            });
        this.logger.debug(() => `Handling root level transforms`);
        const rootLevelTransformGroups = utils.groupTransforms(transforms || []);
        if (rootLevelTransformGroups.wrapTransforms.length > 0) {
            unifiedSchema = wrap.wrapSchema({
                schema: unifiedSchema,
                batch: true,
                transforms,
            });
        }
        else {
            unifiedSchema = utils.applySchemaTransforms(unifiedSchema, undefined, unifiedSchema, rootLevelTransformGroups.noWrapTransforms);
        }
        this.logger.debug(() => `Attaching sources to the unified schema`);
        unifiedSchema.extensions = unifiedSchema.extensions || {};
        Object.defineProperty(unifiedSchema.extensions, 'sourceMap', {
            get: () => sourceMap,
        });
        return unifiedSchema;
    }
}

module.exports = BareMerger;
