// @envelop/live-query/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const graphql = require('graphql');
const graphqlLiveQuery = require('@n1ru4l/graphql-live-query');
const utils = require('@graphql-tools/utils');

const GraphQLLiveDirectiveAST = utils.astFromDirective(graphqlLiveQuery.GraphQLLiveDirective);
const GraphQLLiveDirectiveSDL = graphql.print(GraphQLLiveDirectiveAST);
const useLiveQuery = (opts) => {
    return {
        onExecute: ({ executeFn, setExecuteFn }) => {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore  execute typings do not include AsyncIterable return right now
            setExecuteFn(opts.liveQueryStore.makeExecute(executeFn));
        },
        onValidate: ({ addValidationRule }) => {
            addValidationRule(graphqlLiveQuery.NoLiveMixedWithDeferStreamRule);
        },
        onContextBuilding: ({ extendContext }) => {
            extendContext({
                liveQueryStore: opts.liveQueryStore,
            });
        },
    };
};

Object.defineProperty(exports, 'GraphQLLiveDirective', {
    enumerable: true,
    get: function () {
        return graphqlLiveQuery.GraphQLLiveDirective;
    }
});
exports.GraphQLLiveDirectiveAST = GraphQLLiveDirectiveAST;
exports.GraphQLLiveDirectiveSDL = GraphQLLiveDirectiveSDL;
exports.useLiveQuery = useLiveQuery;
