// @envelop/live-query/index.mjs
import { print } from 'graphql';
import { GraphQLLiveDirective, NoLiveMixedWithDeferStreamRule } from '@n1ru4l/graphql-live-query';
export { GraphQLLiveDirective } from '@n1ru4l/graphql-live-query';
import { astFromDirective } from '@graphql-tools/utils';

const GraphQLLiveDirectiveAST = astFromDirective(GraphQLLiveDirective);
const GraphQLLiveDirectiveSDL = print(GraphQLLiveDirectiveAST);
const useLiveQuery = (opts) => {
    return {
        onExecute: ({ executeFn, setExecuteFn }) => {
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore  execute typings do not include AsyncIterable return right now
            setExecuteFn(opts.liveQueryStore.makeExecute(executeFn));
        },
        onValidate: ({ addValidationRule }) => {
            addValidationRule(NoLiveMixedWithDeferStreamRule);
        },
        onContextBuilding: ({ extendContext }) => {
            extendContext({
                liveQueryStore: opts.liveQueryStore,
            });
        },
    };
};

export { GraphQLLiveDirectiveAST, GraphQLLiveDirectiveSDL, useLiveQuery };
